content/news/2022-09-02-gsoc-final-blog-david-chocholaty.md

```diff
@@ -0,0 +1,714 @@
+avoids dropouts, but the input data samples have to be prefilled
+before the processing, and a delay will be produced before the output
+is produced.  To compare the differences in group delay between the “push model”
+and “pull model” implementation, a measurement was made.
+The results are captured in the following two graphs.
+
+Based on the first graph, the “pull model” implementation with the size
+of the circular buffer of 4096 frames produces a lower latency than
+the “push model” even with the latency produced before the output data.
+Simultaneously, the delay between dynamic Pitch changes is in a smaller range.
+
+*Push and pull model group delay comparison*
+![Pitch Shift effect graph - Push and pull model group delay comparison](
+ {static}/images/news/pitch-shift-effect-graph-1.svg)
+
+In the second graph, the group delay for the unchanged pitch was measured.
+The differences in the measurements are, that the previous pitch values
+before the unchanged pitch settings are different. With the results
+of the last measurement, the “pull model” implementation is more stable
+and produces a similar, almost constant delay after dynamic pitch changes
+for the same pitch setting.
+
+*Push and pull model group delay comparison for unchanged pitch*
+
+![Pitch Shift effect graph - Push and pull model group delay comparison
+for unchanged pitch]({static}/images/news/pitch-shift-effect-graph-2.svg)
+
+The results of the measurements clearly show, that despite the “pull model”
+implementation for the Pitch shift effect in the effect chain is not optimal,
+it should be preferred over the “push model” implementation.
+
+#### Testing and benchmarking
+With the implementation of the `EngineEffectsDelay` for group delay handling
+and the `RingDelayBuffer` as an optimized data structure for the same use case,
+the tests were included with the use of the GoogleTest framework. Basically,
+the common situations were tested, then extreme cases and cases
+that are not allowed but have to be handled for the release builds.
+On the basis that these implemented structures are critical
+from the point of view of performance, with tests, the benchmarks were created
+with the use of the Google Benchmark. Based on the results of benchmarks,
+the used functions and algorithms were compared. After both of the mentioned
+structures were implemented, tested and optimized, the `RingDelayBuffer`
+was introduced in the `EngineEffectsDelay` as an inner structure
+for the group delay handling. Based on the changes and use of optimized
+data structure the performance has highly improved. The performance differences
+are shown in the following benchmarks results taken over
+from the Ubuntu GitHub CI’s results.
+
+##### *Without RingDelayBuffer*
+
+Run on (2 X 2593.91 MHz CPU s)  
+CPU Caches:
+
+* L1 Data 32 KiB (x2)
+* L1 Instruction 32 KiB (x2)
+* L2 Unified 1024 KiB (x2)
+* L3 Unified 36608 KiB (x1)
+
+Load Average: 1.39, 1.82, 1.75
```

Comment:
To be honest, I had to do some research about the meaning :smile:. In a nutshell, it describes the computational work of the CPU. More information in the following sources:

https://en.wikipedia.org/wiki/Load_(computing)
https://www.guyrutenberg.com/2008/07/25/understanding-load-average-a-practitioner-guide/

IMO because the pattern is written exactly the same on the Linux system (the `uptime` command), I propose to just keep it as it is for the user who is consecration in it (others will just probably skip this part) or remove it if we agree that it doesn't make sense to publish. Because it is exactly the copy-paste from the Google Benchmark output, I would prefer to keep it as is. IMHO I don't see much value in explaining the reason for the resulting output.

What's your opinion?

Reply:
Interesting, I didn't know google benchmark outputted this info. To be honest, I'd remove them because they just signify load averages over the last 1, 5, 15 minutes. For a benchmark that only runs for 30s, I don't see much value in the info retrieved with the mentioned window sizes.

Reply:
Sounds reasonable, I agree.